File: scenarios/moderate/mobile/flutter/lib/search_with_silent_updates.dart
import 'package:flutter/material.dart';

class SearchWithSilentUpdates extends StatefulWidget {
  const SearchWithSilentUpdates({Key? key}) : super(key: key);

  @override
  State<SearchWithSilentUpdates> createState() => _SearchWithSilentUpdatesState();
}

class _SearchWithSilentUpdatesState extends State<SearchWithSilentUpdates> {
  final TextEditingController _searchController = TextEditingController();
  List<SearchResult> _suggestions = [];
  List<SearchResult> _results = [];
  bool _isLoading = false;
  bool _showSuggestions = false;
  int _activeSuggestionIndex = -1;

  final List<SearchResult> _mockData = [
    SearchResult(id: 1, title: 'React Development Guide', category: 'Books', author: 'John Doe'),
    SearchResult(id: 2, title: 'Vue.js Tutorial', category: 'Books', author: 'Jane Smith'),
    SearchResult(id: 3, title: 'Angular Framework', category: 'Books', author: 'Bob Johnson'),
    SearchResult(id: 4, title: 'JavaScript Fundamentals', category: 'Books', author: 'Alice Brown'),
    SearchResult(id: 5, title: 'React Components', category: 'Tutorials', author: 'Charlie Wilson'),
    SearchResult(id: 6, title: 'Vue Components', category: 'Tutorials', author: 'Diana Lee'),
    SearchResult(id: 7, title: 'Angular Services', category: 'Tutorials', author: 'Eve Davis'),
    SearchResult(id: 8, title: 'JavaScript ES6', category: 'Tutorials', author: 'Frank Miller'),
  ];

  @override
  void initState() {
    super.initState();
    _searchController.addListener(_handleInput);
  }

  @override
  void dispose() {
    _searchController.dispose();
    super.dispose();
  }

  void _handleInput() {
    final query = _searchController.text;
    if (query.length > 1) {
      final filtered = _mockData.where((item) =>
          item.title.toLowerCase().contains(query.toLowerCase()) ||
          item.category.toLowerCase().contains(query.toLowerCase()) ||
          item.author.toLowerCase().contains(query.toLowerCase())).toList();
      
      setState(() {
        _suggestions = filtered.take(5).toList();
        _showSuggestions = true;
        _activeSuggestionIndex = -1;
      });
    } else {
      setState(() {
        _suggestions = [];
        _showSuggestions = false;
        _activeSuggestionIndex = -1;
      });
    }
  }

  void _handleSearch() async {
    setState(() {
      _isLoading = true;
    });

    // Simulate API call
    await Future.delayed(const Duration(seconds: 1));
    
    final query = _searchController.text;
    final filtered = _mockData.where((item) =>
        item.title.toLowerCase().contains(query.toLowerCase()) ||
        item.category.toLowerCase().contains(query.toLowerCase()) ||
        item.author.toLowerCase().contains(query.toLowerCase())).toList();
    
    setState(() {
      _results = filtered;
      _isLoading = false;
      _showSuggestions = false;
      _activeSuggestionIndex = -1;
    });
    
    // MISSING: Screen reader announcement of results
    // Should announce "Found X results" or "No results found"
  }

  void _handleSuggestionClick(SearchResult suggestion) {
    _searchController.text = suggestion.title;
    setState(() {
      _showSuggestions = false;
      _activeSuggestionIndex = -1;
    });
    _handleSearch();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: const Color(0xFFF5F5F5),
      body: SingleChildScrollView(
        padding: const EdgeInsets.all(20),
        child: Column(
          children: [
            // Main Content Card
            Card(
              elevation: 4,
              child: Padding(
                padding: const EdgeInsets.all(30),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Text(
                      'Search Interface',
                      style: Theme.of(context).textTheme.headlineMedium?.copyWith(
                        fontWeight: FontWeight.bold,
                        color: const Color(0xFF333333),
                      ),
                    ),
                    const SizedBox(height: 16),
                    Text(
                      'Violation: Search input + suggestions dropdown + results area without announcements.',
                      style: Theme.of(context).textTheme.bodyLarge?.copyWith(
                        color: const Color(0xFF666666),
                      ),
                    ),
                    const SizedBox(height: 30),
                    
                    // Search Input Section
                    Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Text(
                          'Search for content:',
                          style: Theme.of(context).textTheme.titleMedium?.copyWith(
                            fontWeight: FontWeight.bold,
                            color: const Color(0xFF333333),
                          ),
                        ),
                        const SizedBox(height: 8),
                        
                        Stack(
                          children: [
                            TextField(
                              controller: _searchController,
                              decoration: const InputDecoration(
                                hintText: 'Type to search...',
                                border: OutlineInputBorder(),
                                prefixIcon: Icon(Icons.search),
                              ),
                              onSubmitted: (_) => _handleSearch(),
                              textInputAction: TextInputAction.search,
                            ),
                            
                            // Suggestions Dropdown - MISSING ARIA ATTRIBUTES
                            if (_showSuggestions && _suggestions.isNotEmpty)
                              Positioned(
                                top: 60,
                                left: 0,
                                right: 0,
                                child: Card(
                                  elevation: 8,
                                  child: Column(
                                    children: _suggestions.asMap().entries.map((entry) {
                                      final index = entry.key;
                                      final suggestion = entry.value;
                                      return InkWell(
                                        onTap: () => _handleSuggestionClick(suggestion),
                                        child: Container(
                                          width: double.infinity,
                                          padding: const EdgeInsets.all(12),
                                          decoration: BoxDecoration(
                                            color: index == _activeSuggestionIndex 
                                                ? const Color(0xFFF8F9FA) 
                                                : Colors.transparent,
                                          ),
                                          child: Column(
                                            crossAxisAlignment: CrossAxisAlignment.start,
                                            children: [
                                              Text(
                                                suggestion.title,
                                                style: const TextStyle(
                                                  fontWeight: FontWeight.bold,
                                                  color: Color(0xFF333333),
                                                ),
                                              ),
                                              Text(
                                                '${suggestion.category} • ${suggestion.author}',
                                                style: const TextStyle(
                                                  fontSize: 12,
                                                  color: Color(0xFF666666),
                                                ),
                                              ),
                                            ],
                                          ),
                                        ),
                                      );
                                    }).toList(),
                                  ),
                                ),
                              ),
                          ],
                        ),
                        
                        const SizedBox(height: 20),
                        
                        SizedBox(
                          width: double.infinity,
                          child: ElevatedButton(
                            onPressed: _isLoading || _searchController.text.trim().isEmpty 
                                ? null 
                                : _handleSearch,
                            child: _isLoading
                                ? const Row(
                                    mainAxisAlignment: MainAxisAlignment.center,
                                    children: [
                                      SizedBox(
                                        width: 20,
                                        height: 20,
                                        child: CircularProgressIndicator(
                                          strokeWidth: 2,
                                          valueColor: AlwaysStoppedAnimation<Color>(Colors.white),
                                        ),
                                      ),
                                      SizedBox(width: 8),
                                      Text('Searching...'),
                                    ],
                                  )
                                : const Text('Search'),
                          ),
                        ),
                      ],
                    ),
                    
                    const SizedBox(height: 30),
                    
                    // Results Section - MISSING LIVE REGION
                    Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Text(
                          'Search Results',
                          style: Theme.of(context).textTheme.titleLarge?.copyWith(
                            fontWeight: FontWeight.bold,
                            color: const Color(0xFF333333),
                          ),
                        ),
                        const SizedBox(height: 16),
                        
                        if (_isLoading)
                          const Center(
                            child: Column(
                              children: [
                                CircularProgressIndicator(),
                                SizedBox(height: 16),
                                Text(
                                  'Searching...',
                                  style: TextStyle(
                                    fontSize: 16,
                                    color: Color(0xFF666666),
                                  ),
                                ),
                              ],
                            ),
                          )
                        else if (_results.isNotEmpty)
                          Column(
                            crossAxisAlignment: CrossAxisAlignment.start,
                            children: [
                              Text(
                                'Found ${_results.length} result${_results.length == 1 ? '' : 's'}',
                                style: const TextStyle(
                                  fontSize: 14,
                                  color: Color(0xFF666666),
                                ),
                              ),
                              const SizedBox(height: 16),
                              SizedBox(
                                height: 300,
                                child: ListView.builder(
                                  itemCount: _results.length,
                                  itemBuilder: (context, index) {
                                    final result = _results[index];
                                    return Card(
                                      margin: const EdgeInsets.only(bottom: 12),
                                      child: Padding(
                                        padding: const EdgeInsets.all(16),
                                        child: Column(
                                          crossAxisAlignment: CrossAxisAlignment.start,
                                          children: [
                                            Text(
                                              result.title,
                                              style: const TextStyle(
                                                fontSize: 16,
                                                fontWeight: FontWeight.w600,
                                                color: Color(0xFF333333),
                                              ),
                                            ),
                                            const SizedBox(height: 8),
                                            Text(
                                              'Category: ${result.category}',
                                              style: const TextStyle(
                                                fontSize: 12,
                                                color: Color(0xFF666666),
                                              ),
                                            ),
                                            Text(
                                              'Author: ${result.author}',
                                              style: const TextStyle(
                                                fontSize: 12,
                                                color: Color(0xFF666666),
                                              ),
                                            ),
                                          ],
                                        ),
                                      ),
                                    );
                                  },
                                ),
                              ),
                            ],
                          )
                        else if (_searchController.text.isNotEmpty)
                          const Center(
                            child: Text(
                              'No results found',
                              style: TextStyle(
                                fontSize: 16,
                                color: Color(0xFF666666),
                              ),
                            ),
                          )
                        else
                          const Center(
                            child: Text(
                              'Enter a search term to find content',
                              style: TextStyle(
                                fontSize: 16,
                                color: Color(0xFF666666),
                              ),
                            ),
                          ),
                      ],
                    ),
                  ],
                ),
              ),
            ),
            
            const SizedBox(height: 20),
            
            // Accessibility Information Card
            Card(
              color: const Color(0xFFE9ECEF),
              child: Padding(
                padding: const EdgeInsets.all(20),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Text(
                      'Accessibility Issues:',
                      style: Theme.of(context).textTheme.titleMedium?.copyWith(
                        fontWeight: FontWeight.bold,
                        color: const Color(0xFF333333),
                      ),
                    ),
                    const SizedBox(height: 12),
                    const Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Text('• Missing live region: No announcements when search results change'),
                        Text('• Incomplete ARIA attributes: Suggestions lack proper aria-activedescendant'),
                        Text('• No keyboard navigation: Arrow keys don\'t navigate through suggestions'),
                        Text('• Missing announcements: Screen readers don\'t announce result count changes'),
                        Text('• No loading state announcement: Loading state not announced to screen readers'),
                        Text('• Missing suggestion selection: No aria-selected state management'),
                      ],
                    ),
                    const SizedBox(height: 20),
                    Text(
                      'How to Fix:',
                      style: Theme.of(context).textTheme.titleMedium?.copyWith(
                        fontWeight: FontWeight.bold,
                        color: const Color(0xFF333333),
                      ),
                    ),
                    const SizedBox(height: 12),
                    const Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Text('• Add aria-live="polite" region for result announcements'),
                        Text('• Implement aria-activedescendant for suggestion navigation'),
                        Text('• Add arrow key navigation (Up/Down) through suggestions'),
                        Text('• Announce result count changes ("Found X results")'),
                        Text('• Add loading state announcements'),
                        Text('• Implement proper aria-selected state management'),
                      ],
                    ),
                  ],
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class SearchResult {
  final int id;
  final String title;
  final String category;
  final String author;

  SearchResult({
    required this.id,
    required this.title,
    required this.category,
    required this.author,
  });
}

File: scenarios/moderate/mobile/ios/swiftui/SearchWithSilentUpdates.swift
import SwiftUI

struct SearchWithSilentUpdates: View {
    @State private var query = ""
    @State private var suggestions: [SearchResult] = []
    @State private var results: [SearchResult] = []
    @State private var isLoading = false
    @State private var showSuggestions = false
    @State private var activeSuggestionIndex = -1
    @FocusState private var isSearchFocused: Bool
    
    let mockData = [
        SearchResult(id: 1, title: "React Development Guide", category: "Books", author: "John Doe"),
        SearchResult(id: 2, title: "Vue.js Tutorial", category: "Books", author: "Jane Smith"),
        SearchResult(id: 3, title: "Angular Framework", category: "Books", author: "Bob Johnson"),
        SearchResult(id: 4, title: "JavaScript Fundamentals", category: "Books", author: "Alice Brown"),
        SearchResult(id: 5, title: "React Components", category: "Tutorials", author: "Charlie Wilson"),
        SearchResult(id: 6, title: "Vue Components", category: "Tutorials", author: "Diana Lee"),
        SearchResult(id: 7, title: "Angular Services", category: "Tutorials", author: "Eve Davis"),
        SearchResult(id: 8, title: "JavaScript ES6", category: "Tutorials", author: "Frank Miller")
    ]
    
    var body: some View {
        VStack(spacing: 20) {
            VStack(alignment: .leading, spacing: 16) {
                Text("Search Interface")
                    .font(.title)
                    .fontWeight(.bold)
                
                Text("Search for content using the input field below. Suggestions will appear as you type.")
                    .font(.body)
                    .foregroundColor(.secondary)
                
                // Search Input Section
                VStack(alignment: .leading, spacing: 8) {
                    Text("Search for content:")
                        .font(.headline)
                        .fontWeight(.bold)
                    
                    VStack(alignment: .leading, spacing: 0) {
                        TextField("Type to search...", text: $query)
                            .textFieldStyle(RoundedBorderTextFieldStyle())
                            .focused($isSearchFocused)
                            .onChange(of: query) { _ in
                                handleInput()
                            }
                            .accessibilityLabel("Search input")
                            .accessibilityHint("Enter search terms to find content")
                            .accessibilityValue(query.isEmpty ? "Empty" : query)
                        
                        // Suggestions Dropdown - MISSING ARIA ATTRIBUTES
                        if showSuggestions && !suggestions.isEmpty {
                            VStack(spacing: 0) {
                                ForEach(Array(suggestions.enumerated()), id: \.offset) { index, suggestion in
                                    Button(action: {
                                        handleSuggestionClick(suggestion)
                                    }) {
                                        VStack(alignment: .leading, spacing: 4) {
                                            Text(suggestion.title)
                                                .font(.headline)
                                                .foregroundColor(.primary)
                                            Text("\(suggestion.category) • \(suggestion.author)")
                                                .font(.caption)
                                                .foregroundColor(.secondary)
                                        }
                                        .frame(maxWidth: .infinity, alignment: .leading)
                                        .padding()
                                        .background(index == activeSuggestionIndex ? Color(.systemGray5) : Color.clear)
                                    }
                                    .buttonStyle(PlainButtonStyle())
                                    .accessibilityLabel(suggestion.title)
                                    .accessibilityHint("Tap to select this suggestion")
                                    
                                    if index < suggestions.count - 1 {
                                        Divider()
                                    }
                                }
                            }
                            .background(Color(.systemBackground))
                            .overlay(
                                RoundedRectangle(cornerRadius: 8)
                                    .stroke(Color(.systemGray4), lineWidth: 1)
                            )
                            .cornerRadius(8)
                            .shadow(radius: 4)
                        }
                    }
                    
                    Button(action: handleSearch) {
                        HStack {
                            if isLoading {
                                ProgressView()
                                    .scaleEffect(0.8)
                            }
                            Text(isLoading ? "Searching..." : "Search")
                        }
                        .frame(maxWidth: .infinity)
                        .padding()
                        .background(isLoading ? Color.gray : Color.blue)
                        .foregroundColor(.white)
                        .cornerRadius(8)
                    }
                    .disabled(isLoading || query.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
                    .accessibilityLabel("Search button")
                    .accessibilityHint("Tap to search for content")
                }
                
                // Results Section - MISSING LIVE REGION
                VStack(alignment: .leading, spacing: 16) {
                    Text("Search Results")
                        .font(.title2)
                        .fontWeight(.bold)
                    
                    if isLoading {
                        VStack {
                            ProgressView()
                            Text("Searching...")
                                .foregroundColor(.secondary)
                        }
                        .frame(maxWidth: .infinity)
                        .padding()
                    } else if !results.isEmpty {
                        VStack(alignment: .leading, spacing: 12) {
                            Text("Found \(results.count) result\(results.count == 1 ? "" : "s")")
                                .font(.caption)
                                .foregroundColor(.secondary)
                            
                            LazyVStack(spacing: 12) {
                                ForEach(results) { result in
                                    VStack(alignment: .leading, spacing: 8) {
                                        Text(result.title)
                                            .font(.headline)
                                            .fontWeight(.semibold)
                                        Text("Category: \(result.category)")
                                            .font(.caption)
                                            .foregroundColor(.secondary)
                                        Text("Author: \(result.author)")
                                            .font(.caption)
                                            .foregroundColor(.secondary)
                                    }
                                    .frame(maxWidth: .infinity, alignment: .leading)
                                    .padding()
                                    .background(Color(.systemGray6))
                                    .cornerRadius(8)
                                }
                            }
                        }
                    } else if !query.isEmpty {
                        VStack {
                            Text("No results found for \"\(query)\"")
                                .foregroundColor(.secondary)
                        }
                        .frame(maxWidth: .infinity)
                        .padding()
                    } else {
                        VStack {
                            Text("Enter a search term to find content")
                                .foregroundColor(.secondary)
                        }
                        .frame(maxWidth: .infinity)
                        .padding()
                    }
                }
            }
            .padding()
            .background(Color(.systemBackground))
            .cornerRadius(12)
            .shadow(radius: 4)
            
            // Accessibility Information
            VStack(alignment: .leading, spacing: 12) {
                Text("Accessibility Issues:")
                    .font(.headline)
                    .fontWeight(.bold)
                
                VStack(alignment: .leading, spacing: 8) {
                    accessibilityIssue("Missing live region: No announcements when search results change")
                    accessibilityIssue("Incomplete ARIA attributes: Suggestions lack proper aria-activedescendant")
                    accessibilityIssue("No keyboard navigation: Arrow keys don't navigate through suggestions")
                    accessibilityIssue("Missing announcements: Screen readers don't announce result count changes")
                    accessibilityIssue("No loading state announcement: Loading state not announced to screen readers")
                    accessibilityIssue("Missing suggestion selection: No aria-selected state management")
                }
                
                Text("How to Fix:")
                    .font(.headline)
                    .fontWeight(.bold)
                    .padding(.top)
                
                VStack(alignment: .leading, spacing: 8) {
                    fixItem("Add aria-live=\"polite\" region for result announcements")
                    fixItem("Implement aria-activedescendant for suggestion navigation")
                    fixItem("Add arrow key navigation (Up/Down) through suggestions")
                    fixItem("Announce result count changes (\"Found X results\")")
                    fixItem("Add loading state announcements")
                    fixItem("Implement proper aria-selected state management")
                }
            }
            .padding()
            .background(Color(.systemGray6))
            .cornerRadius(8)
        }
        .padding()
        .background(Color(.systemGroupedBackground))
        .onTapGesture {
            if showSuggestions {
                showSuggestions = false
                activeSuggestionIndex = -1
            }
        }
    }
    
    private func handleInput() {
        if query.count > 1 {
            let filtered = mockData.filter { item in
                item.title.lowercased().contains(query.lowercased()) ||
                item.category.lowercased().contains(query.lowercased()) ||
                item.author.lowercased().contains(query.lowercased())
            }
            suggestions = Array(filtered.prefix(5))
            showSuggestions = true
            activeSuggestionIndex = -1
        } else {
            suggestions = []
            showSuggestions = false
            activeSuggestionIndex = -1
        }
    }
    
    private func handleSearch() {
        isLoading = true
        
        // Simulate API call
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
            let filtered = mockData.filter { item in
                item.title.lowercased().contains(query.lowercased()) ||
                item.category.lowercased().contains(query.lowercased()) ||
                item.author.lowercased().contains(query.lowercased())
            }
            results = filtered
            isLoading = false
            showSuggestions = false
            activeSuggestionIndex = -1
            
            // MISSING: Screen reader announcement of results
            // Should announce "Found X results" or "No results found"
        }
    }
    
    private func handleSuggestionClick(_ suggestion: SearchResult) {
        query = suggestion.title
        showSuggestions = false
        activeSuggestionIndex = -1
        handleSearch()
    }
    
    @ViewBuilder
    private func accessibilityIssue(_ text: String) -> some View {
        HStack(alignment: .top, spacing: 8) {
            Text("•")
                .fontWeight(.bold)
            Text(text)
                .font(.body)
        }
    }
    
    @ViewBuilder
    private func fixItem(_ text: String) -> some View {
        HStack(alignment: .top, spacing: 8) {
            Text("•")
                .fontWeight(.bold)
            Text(text)
                .font(.body)
        }
    }
}

struct SearchResult: Identifiable {
    let id: Int
    let title: String
    let category: String
    let author: String
}

// MARK: - Preview
struct SearchWithSilentUpdates_Previews: PreviewProvider {
    static var previews: some View {
        SearchWithSilentUpdates()
            .previewDisplayName("Search With Silent Updates")
    }
}